src/Mod/Sketcher/Gui/TaskSketcherElements.cpp

```diff
@@ -233,21 +371,128 @@ void ElementView::deleteSelectedItems()
+
+ElementItemDelegate::ElementItemDelegate(QObject* parent) : QStyledItemDelegate(parent)
 {
-    switch (event->key())
-    {
-      case Qt::Key_Z:
-        // signal
-        Q_EMIT onFilterShortcutPressed();
-        break;
-      default:
-        QListWidget::keyPressEvent( event );
-        break;
+}
+
+ElementItemDelegate::~ElementItemDelegate()
+{
+}
+
+
+void ElementItemDelegate::paint(QPainter* painter, const QStyleOptionViewItem& option, const QModelIndex& index) const
+{
+    ElementItem* item = getElementtItem(index);
+
+    if (item) {
+
+        int border = 1; //1px, looks good around buttons.
+        int height = option.rect.height();
+        int x0 = option.rect.x() + 4; //4px on the left of icons, looks good.
+        int iconsize = height - 2 * border;
+        int btny = option.rect.y() + border;
+
+        if (item->isLineSelected || item->isStartingPointSelected || item->isEndPointSelected || item->isMidPointSelected) { //option.state & QStyle::State_Selected
+            painter->fillRect(option.rect, option.palette.highlight());
+            if (!item->isLineSelected && item->GeometryType != Part::GeomPoint::getClassTypeId()) {
+                QRect rect0(x0 + border, btny, iconsize, iconsize);
+                painter->fillRect(rect0, option.palette.base());
+            }
+            if (!item->isStartingPointSelected && item->GeometryType != Part::GeomCircle::getClassTypeId() && item->GeometryType != Part::GeomEllipse::getClassTypeId()) {
+                QRect rect1(x0 + height + border, btny, iconsize, iconsize);
+                painter->fillRect(rect1, option.palette.base());
+            }
+            if (!item->isEndPointSelected && item->GeometryType != Part::GeomCircle::getClassTypeId() && item->GeometryType != Part::GeomEllipse::getClassTypeId() && item->GeometryType != Part::GeomPoint::getClassTypeId()) {
+                QRect rect2(x0 + height * 2 + border, btny, iconsize, iconsize);
+                painter->fillRect(rect2, option.palette.base());
+            }
+            if (!item->isMidPointSelected && item->GeometryType != Part::GeomLineSegment::getClassTypeId() && item->GeometryType != Part::GeomBSplineCurve::getClassTypeId() && item->GeometryType != Part::GeomPoint::getClassTypeId()){
+                QRect rect3(x0 + height * 3 + border, btny, iconsize, iconsize);
+                painter->fillRect(rect3, option.palette.base());
+            }
+        }
+
+        painter->drawPixmap(x0 + border             , btny, item->icon0.pixmap(iconsize, iconsize));
+        painter->drawPixmap(x0 + border + height    , btny, item->icon1.pixmap(iconsize, iconsize));
+        painter->drawPixmap(x0 + border + height * 2, btny, item->icon2.pixmap(iconsize, iconsize));
+        painter->drawPixmap(x0 + border + height * 3, btny, item->icon3.pixmap(iconsize, iconsize));
+
+        //Label :
+        painter->drawText(x0 + height * 4 + 3 * border, option.rect.y() + height - 5, item->label);

```

Comment:
Because the icons are square. So each box holding one icon is a square of height and width = item.height(). So positioning of boxes use height for each icon. The second box starts at 1 height from x = 0. and so on

Reply:
My opinion is that it may change in the future and will be a mess to maintain. Worth putting `int width = height` when initializing variables, then use `width` where it talks about ... width. ;)

Reply:
ok